## Primera exploración


In [1]:
# Tratamiento de datos
import numpy as np
import pandas as pd

# Gráficos
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go

# Estadísticas
import scipy
 
# mis cosas
import sys
import os

# Añado el directorio padre (del que está este notebook) a sys.path
sys.path.append(os.path.abspath('../'))
from scripts.utils_agv import ini_inspec


(primeramente renombramos el Dataframe de trabajo, a la vez que hacemos una copia para no tener que cargarlo de nuevo si lo "estropeamos")

In [2]:
viviendas_Madrid = pd.read_csv('../data/raw/idealista Madrid.csv')
df = viviendas_Madrid.copy()

In [3]:
df.head(2)

url  \
0  https://www.idealista.com/inmueble/104027174/   
1  https://www.idealista.com/inmueble/102321942/   

                                          listingUrl  \
0  https://www.idealista.com/venta-viviendas/madr...   
1  https://www.idealista.com/venta-viviendas/madr...   

                                       title         id    price  baths  \
0       Piso en venta en calle de Villanueva  104027174  1920000      3   
1  Piso en venta en calle de Núñez de Balboa  102321942  1995000      3   

   rooms  sqft                                        description  \
0      3   183  Residencia única con acabados de la más alta c...   
1      3   170  Preciosa reforma a estrenar, con terrazas y ga...   

              address typology advertiserProfessionalName  \
0   Recoletos, Madrid    Pisos            Promora Madrid    
1  Castellana, Madrid    Pisos                Madrid MMC    

           advertiserName  
0          Promora Madrid  
1  Engel & Völkers Madrid

In [4]:
ini_inspec(df)

=== TAMAÑO Y ESTRUCTURA DE LOS DATOS ===
Número total de registros (filas): 915
Número de columnas: 13
Uso de memoria: 93.06 KB


=== TIPOS DE DATOS Y NOMBRES DE COLUMNAS ===
url                           object
listingUrl                    object
title                         object
id                             int64
price                          int64
baths                          int64
rooms                          int64
sqft                           int64
description                   object
address                       object
typology                      object
advertiserProfessionalName    object
advertiserName                object
dtype: object


Información detallada del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 915 entries, 0 to 914
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   url                         915 non-null    object
 1   listingUrl   

In [5]:
print(df['typology'].unique())
print(df['address'].unique())

#de momento, no parece interesante la cardinalidad del anunciante para hacer hipotesis
#print(df['advertiserProfessionalName'].unique())
#print(df['advertiserName'].unique())

['Pisos' 'Independientes']
['Recoletos, Madrid' 'Castellana, Madrid' 'Hortaleza, Madrid'
 'Chamartín, Madrid' 'Cuatro Caminos, Madrid' 'Goya, Madrid'
 'Aravaca, Madrid' 'Sol, Madrid' 'Bernabéu-Hispanoamérica, Madrid'
 'Vallehermoso, Madrid' 'Huertas-Cortes, Madrid' 'Chamberí, Madrid'
 'Nueva España, Madrid' 'Conde Orgaz-Piovera, Madrid'
 'Chueca-Justicia, Madrid' 'Argüelles, Madrid'
 'Nuevos Ministerios-Ríos Rosas, Madrid' 'Centro, Madrid'
 'Moncloa, Madrid' 'Retiro, Madrid' 'Barrio de Salamanca, Madrid'
 'Peñagrande, Madrid' 'Mirasierra, Madrid' 'Fuencarral, Madrid'
 'Comillas, Madrid' 'Malasaña-Universidad, Madrid'
 'Palos de Moguer, Madrid' 'San Blas, Madrid' 'Prosperidad, Madrid'
 'Bellas Vistas, Madrid' 'Niño Jesús, Madrid' 'El Viso, Madrid'
 'Palacio, Madrid' 'Lista, Madrid' 'Ciudad Lineal, Madrid'
 'Lavapiés-Embajadores, Madrid' 'Salvador, Madrid' 'Almagro, Madrid'
 'Fuente del Berro, Madrid' 'Gaztambide, Madrid' 'Acacias, Madrid'
 'Arapiles, Madrid' 'Guindalera, Madrid' 'Ciudad

Veamos el conteo de cuantos unicos hay:

In [6]:
print(len(df['address'].unique())) 
print(df['address'].nunique())

105
105


Veamos si todos los únicos contienen "Madrid", o alguno es de otra ciudad:

In [7]:
pd.Series(df['address'].unique()).str.contains('Madrid')
#len(pd.Series(df['address'].unique()).str.contains('Madrid'))

0      True
1      True
2      True
3      True
4      True
       ... 
100    True
101    True
102    True
103    True
104    True
Length: 105, dtype: bool

Confirmamos que todas las direcciones corresponden a Madrid. 

Voy a hacer una investigación de distritos y barrios con un par de tablas que he descargado de
https://datos.gob.es/en/catalogo?publisher_display_name=Ayuntamiento+de+Madrid





In [8]:
distritos_Madrid = pd.read_csv('../data/external/Madrid_Distritos.csv', sep=';')
distritos_Madrid


OBJECTID   Shape_Leng   Shape_Area  COD_DIS  COD_DIS_TX  \
0         11  10304,07973  5228246,741        1           1   
1          6  12806,76426   6462175,63        2           2   
2          9  9523,885329  5466211,164        3           3   
3         12  10866,33577  5392403,428        4           4   
4         17  13396,81708  9175481,878        5           5   
5         15  9919,059595  5374724,446        6           6   
6         13  9019,542776  4679185,047        7           7   
7         21  94482,14556  237838370,4        8           8   
8         16  41034,34336   46531136,2        9           9   
9          7  28181,05527   25427202,6       10          10   
10         4  18083,30558  14048316,11       11          11   
11         3  11878,91045  7777717,293       12          12   
12         5  17263,77382  14968606,15       13          13   
13         8  10814,07229  6103167,529       14          14   
14        18  21524,70026  11425677,82       15          15   
15        20  26988,38785  27419761,67       16          16   
16         1  18342,36139  20187586,34       17          17   
17         2  33480,66228  51467238,51       18          18   
18        10  34265,39005  35266713,27       19          19   
19        14  26780,32886  22292392,33       20          20   
20        19  28730,35067  41922811,47       21          21   

               DISTRI_MAY                 NOMBRE              DISTRI_MT  
0                  CENTRO                 Centro                 CENTRO  
1              ARGANZUELA             Arganzuela             ARGANZUELA  
2                  RETIRO                 Retiro                 RETIRO  
3               SALAMANCA              Salamanca              SALAMANCA  
4               CHAMARTIN              Chamartín              CHAMARTÍN  
5                  TETUAN                 Tetuán                 TETUÁN  
6                CHAMBERI               Chamberí               CHAMBERÍ  
7   FUENCARRAL - EL PARDO  Fuencarral - El Pardo  FUENCARRAL - EL PARDO  
8       MONCLOA - ARAVACA      Moncloa - Aravaca      MONCLOA - ARAVACA  
9                  LATINA                 Latina                 LATINA  
10            CARABANCHEL            Carabanchel            CARABANCHEL  
11                  USERA                  Usera                  USERA  
12     PUENTE DE VALLECAS     Puente de Vallecas     PUENTE DE VALLECAS  
13              MORATALAZ              Moratalaz              MORATALAZ  
14          CIUDAD LINEAL          Ciudad Lineal          CIUDAD LINEAL  
15              HORTALEZA              Hortaleza              HORTALEZA  
16             VILLAVERDE             Villaverde             VILLAVERDE  
17      VILLA DE VALLECAS      Villa de Vallecas      VILLA DE VALLECAS  
18              VICALVARO              Vicálvaro              VICÁLVARO  
19  SAN BLAS - CANILLEJAS  San Blas - Canillejas  SAN BLAS - CANILLEJAS  
20                BARAJAS                Barajas                BARAJAS

In [9]:
dM = distritos_Madrid.copy().set_index('OBJECTID')
dM

Shape_Leng   Shape_Area  COD_DIS  COD_DIS_TX  \
OBJECTID                                                  
11        10304,07973  5228246,741        1           1   
6         12806,76426   6462175,63        2           2   
9         9523,885329  5466211,164        3           3   
12        10866,33577  5392403,428        4           4   
17        13396,81708  9175481,878        5           5   
15        9919,059595  5374724,446        6           6   
13        9019,542776  4679185,047        7           7   
21        94482,14556  237838370,4        8           8   
16        41034,34336   46531136,2        9           9   
7         28181,05527   25427202,6       10          10   
4         18083,30558  14048316,11       11          11   
3         11878,91045  7777717,293       12          12   
5         17263,77382  14968606,15       13          13   
8         10814,07229  6103167,529       14          14   
18        21524,70026  11425677,82       15          15   
20        26988,38785  27419761,67       16          16   
1         18342,36139  20187586,34       17          17   
2         33480,66228  51467238,51       18          18   
10        34265,39005  35266713,27       19          19   
14        26780,32886  22292392,33       20          20   
19        28730,35067  41922811,47       21          21   

                     DISTRI_MAY                 NOMBRE              DISTRI_MT  
OBJECTID                                                                       
11                       CENTRO                 Centro                 CENTRO  
6                    ARGANZUELA             Arganzuela             ARGANZUELA  
9                        RETIRO                 Retiro                 RETIRO  
12                    SALAMANCA              Salamanca              SALAMANCA  
17                    CHAMARTIN              Chamartín              CHAMARTÍN  
15                       TETUAN                 Tetuán                 TETUÁN  
13                     CHAMBERI               Chamberí               CHAMBERÍ  
21        FUENCARRAL - EL PARDO  Fuencarral - El Pardo  FUENCARRAL - EL PARDO  
16            MONCLOA - ARAVACA      Moncloa - Aravaca      MONCLOA - ARAVACA  
7                        LATINA                 Latina                 LATINA  
4                   CARABANCHEL            Carabanchel            CARABANCHEL  
3                         USERA                  Usera                  USERA  
5            PUENTE DE VALLECAS     Puente de Vallecas     PUENTE DE VALLECAS  
8                     MORATALAZ              Moratalaz              MORATALAZ  
18                CIUDAD LINEAL          Ciudad Lineal          CIUDAD LINEAL  
20                    HORTALEZA              Hortaleza              HORTALEZA  
1                    VILLAVERDE             Villaverde             VILLAVERDE  
2             VILLA DE VALLECAS      Villa de Vallecas      VILLA DE VALLECAS  
10                    VICALVARO              Vicálvaro              VICÁLVARO  
14        SAN BLAS - CANILLEJAS  San Blas - Canillejas  SAN BLAS - CANILLEJAS  
19                      BARAJAS                Barajas                BARAJAS

Hasta aquí el DF de los 21 distritos de Madrid, limpiado del indice

In [10]:
distritos_datos = pd.read_csv('../data/external/distritos_municipio_madrid.csv',sep=';')
distritos_datos

distrito_codigo           distrito_nombre  municipio_codigo  \
0             79601                   Centro                796   
1             79602               Arganzuela                796   
2             79603                   Retiro                796   
3             79604                Salamanca                796   
4             79605                Chamartín                796   
5             79606                   Tetuan                796   
6             79607                 Chamberí                796   
7             79608      Fuencarral-El Pardo                796   
8             79609          Moncloa-Aravaca                796   
9             79610                   Latina                796   
10            79611              Carabanchel                796   
11            79612                    Usera                796   
12            79613       Puente de Vallecas                796   
13            79614                Moratalaz                796   
14            79615            Ciudad Lineal                796   
15            79616                Hortaleza                796   
16            79617               Villaverde                796   
17            79618        Villa de Vallecas                796   
18            79619                Vicálvaro                796   
19            79620      San Blas-Canillejas                796   
20            79621                  Barajas                796   

   municipio_nombre  superficie_km2  densidad_por_km2  
0            Madrid            5.21          25340.69  
1            Madrid            6.52          23306.44  
2            Madrid            5.42          21867.53  
3            Madrid            5.36          26830.78  
4            Madrid            9.12          15723.25  
5            Madrid            5.37          28664.25  
6            Madrid            4.73          29049.26  
7            Madrid          238.00           1003.00  
8            Madrid           46.47           2515.26  
9            Madrid           25.47           9183.75  
10           Madrid           14.10          17316.88  
11           Madrid            7.70          17535.32  
12           Madrid           14.84          15345.01  
13           Madrid            6.08          15493.59  
14           Madrid           11.52          18455.56  
15           Madrid           25.87           6973.33  
16           Madrid           20.21           7059.13  
17           Madrid           51.49           2026.82  
18           Madrid           35.36           1981.11  
19           Madrid           22.26           6934.37  
20           Madrid           43.56           1076.06

In [11]:
dMd = distritos_datos.copy()
dMd

distrito_codigo           distrito_nombre  municipio_codigo  \
0             79601                   Centro                796   
1             79602               Arganzuela                796   
2             79603                   Retiro                796   
3             79604                Salamanca                796   
4             79605                Chamartín                796   
5             79606                   Tetuan                796   
6             79607                 Chamberí                796   
7             79608      Fuencarral-El Pardo                796   
8             79609          Moncloa-Aravaca                796   
9             79610                   Latina                796   
10            79611              Carabanchel                796   
11            79612                    Usera                796   
12            79613       Puente de Vallecas                796   
13            79614                Moratalaz                796   
14            79615            Ciudad Lineal                796   
15            79616                Hortaleza                796   
16            79617               Villaverde                796   
17            79618        Villa de Vallecas                796   
18            79619                Vicálvaro                796   
19            79620      San Blas-Canillejas                796   
20            79621                  Barajas                796   

   municipio_nombre  superficie_km2  densidad_por_km2  
0            Madrid            5.21          25340.69  
1            Madrid            6.52          23306.44  
2            Madrid            5.42          21867.53  
3            Madrid            5.36          26830.78  
4            Madrid            9.12          15723.25  
5            Madrid            5.37          28664.25  
6            Madrid            4.73          29049.26  
7            Madrid          238.00           1003.00  
8            Madrid           46.47           2515.26  
9            Madrid           25.47           9183.75  
10           Madrid           14.10          17316.88  
11           Madrid            7.70          17535.32  
12           Madrid           14.84          15345.01  
13           Madrid            6.08          15493.59  
14           Madrid           11.52          18455.56  
15           Madrid           25.87           6973.33  
16           Madrid           20.21           7059.13  
17           Madrid           51.49           2026.82  
18           Madrid           35.36           1981.11  
19           Madrid           22.26           6934.37  
20           Madrid           43.56           1076.06

Y aquí un DF con los datos de densidad, extensión, etc de los 21 distritos de Madrid. Realizada copia renombrado cómodo.


Voy a hacer los mismo con los barrios, que son 131

In [12]:
barrios_Madrid = pd.read_csv('../data/external/Madrid_Barrios.csv', sep=';')
barrios_Madrid

CODDIS   NOMDIS  COD_BAR                      NOMBRE  COD_DIS_TX  \
0         1   Centro       12                 Embajadores           1   
1         1   Centro       13                      Cortes           1   
2         1   Centro       16                         Sol           1   
3         1   Centro       11                     Palacio           1   
4         1   Centro       14                    Justicia           1   
..      ...      ...      ...                         ...         ...   
126      21  Barajas      211            Alameda de Osuna          21   
127      21  Barajas      215                  Corralejos          21   
128      21  Barajas      213  Casco Histórico de Barajas          21   
129      21  Barajas      214                       Timón          21   
130      21  Barajas      212                  Aeropuerto          21   

                     BARRIO_MAY  COD_DISBAR  NUM_BAR  \
0                   EMBAJADORES          12        2   
1                        CORTES          13        3   
2                           SOL          16        6   
3                       PALACIO          11        1   
4                      JUSTICIA          14        4   
..                          ...         ...      ...   
126            ALAMEDA DE OSUNA         211        1   
127                  CORRALEJOS         215        5   
128  CASCO HISTORICO DE BARAJAS         213        3   
129                       TIMON         214        4   
130                  AEROPUERTO         212        2   

                      BARRIO_MT COD_DISB  
0                   EMBAJADORES    1-feb  
1                        CORTES    1-mar  
2                           SOL    1-jun  
3                       PALACIO    1-ene  
4                      JUSTICIA    1-abr  
..                          ...      ...  
126            ALAMEDA DE OSUNA   21-ene  
127                  CORRALEJOS   21-may  
128  CASCO HISTÓRICO DE BARAJAS   21-mar  
129                       TIMÓN   21-abr  
130                  AEROPUERTO   21-feb  

[131 rows x 10 columns]

In [13]:
bM = barrios_Madrid.copy().set_index('CODDIS')
bM.head(45)


NOMDIS  COD_BAR                NOMBRE  COD_DIS_TX  \
CODDIS                                                                     
1                      Centro       12           Embajadores           1   
1                      Centro       13                Cortes           1   
1                      Centro       16                   Sol           1   
1                      Centro       11               Palacio           1   
1                      Centro       14              Justicia           1   
1                      Centro       15           Universidad           1   
2                  Arganzuela       24               Legazpi           2   
2                  Arganzuela       23               Chopera           2   
2                  Arganzuela       25              Delicias           2   
2                  Arganzuela       22               Acacias           2   
2                  Arganzuela       26  Palos de la Frontera           2   
2                  Arganzuela       27                Atocha           2   
2                  Arganzuela       21              Imperial           2   
3                      Retiro       32               Adelfas           3   
3                      Retiro       31              Pacífico           3   
3                      Retiro       36            Niño Jesús           3   
3                      Retiro       33              Estrella           3   
3                      Retiro       35         Los Jerónimos           3   
3                      Retiro       34                 Ibiza           3   
4                   Salamanca       42                  Goya           4   
4                   Salamanca       41             Recoletos           4   
4                   Salamanca       43      Fuente del Berro           4   
4                   Salamanca       45                 Lista           4   
4                   Salamanca       46            Castellana           4   
4                   Salamanca       44            Guindalera           4   
5                   Chamartín       51               El Viso           5   
5                   Chamartín       53         Ciudad Jardín           5   
5                   Chamartín       52           Prosperidad           5   
5                   Chamartín       54        Hispanoamérica           5   
5                   Chamartín       55          Nueva España           5   
5                   Chamartín       56              Castilla           5   
6                      Tetuán       62        Cuatro Caminos           6   
6                      Tetuán       61         Bellas Vistas           6   
6                      Tetuán       66            Berruguete           6   
6                      Tetuán       63           Castillejos           6   
6                      Tetuán       65         Valdeacederas           6   
6                      Tetuán       64              Almenara           6   
7                    Chamberí       74               Almagro           7   
7                    Chamberí       73             Trafalgar           7   
7                    Chamberí       72              Arapiles           7   
7                    Chamberí       71            Gaztambide           7   
7                    Chamberí       75            Ríos Rosas           7   
7                    Chamberí       76          Vallehermoso           7   
8       Fuencarral - El Pardo       84                 Pilar           8   
8       Fuencarral - El Pardo       83            Peñagrande           8   

                  BARRIO_MAY  COD_DISBAR  NUM_BAR             BARRIO_MT  \
CODDIS                                                                    
1                EMBAJADORES          12        2           EMBAJADORES   
1                     CORTES          13        3                CORTES   
1                        SOL          16        6                   SOL   
1                    PALACIO          11        1               PALACIO   


In [14]:
bM.iloc[14,2]  #esta es una comprobación, pues antes estaba mal escrito y tuve que editar el csv

'Pacífico'

In [15]:
barrios_datos = pd.read_csv('../data/external/barrios_municipio_madrid.csv',sep=';',encoding='latin1')
barrios_datos

barrio_codigo                       barrio_nombre  distrito_codigo  \
0           796011                             Palacio            79601   
1           796012                         Embajadores            79601   
2           796013                              Cortes            79601   
3           796014                            Justicia            79601   
4           796015                         Universidad            79601   
..             ...                                 ...              ...   
123         796211                    Alameda de Osuna            79621   
124         796212                          Aeropuerto            79621   
125         796213          Casco Histórico de Barajas            79621   
126         796214                               Timón            79621   
127         796215                          Corralejos            79621   

    distrito_nombre  municipio_codigo municipio_nombre  superficie_km2  \
0           Centro                796           Madrid            1.46   
1           Centro                796           Madrid            1.03   
2           Centro                796           Madrid            0.59   
3           Centro                796           Madrid            0.75   
4           Centro                796           Madrid            0.93   
..              ...               ...              ...             ...   
123        Barajas                796           Madrid            1.98   
124        Barajas                796           Madrid           19.87   
125        Barajas                796           Madrid            0.64   
126        Barajas                796           Madrid           16.40   
127        Barajas                796           Madrid            4.67   

     densidad_por_km2  
0            15323.29  
1            43345.63  
2            17850.85  
3            21866.67  
4            33051.61  
..                ...  
123           9795.45  
124             89.88  
125          11415.62  
126            672.01  
127           1577.09  

[128 rows x 8 columns]

In [16]:
bMd = barrios_datos.copy()
bMd

barrio_codigo                       barrio_nombre  distrito_codigo  \
0           796011                             Palacio            79601   
1           796012                         Embajadores            79601   
2           796013                              Cortes            79601   
3           796014                            Justicia            79601   
4           796015                         Universidad            79601   
..             ...                                 ...              ...   
123         796211                    Alameda de Osuna            79621   
124         796212                          Aeropuerto            79621   
125         796213          Casco Histórico de Barajas            79621   
126         796214                               Timón            79621   
127         796215                          Corralejos            79621   

    distrito_nombre  municipio_codigo municipio_nombre  superficie_km2  \
0           Centro                796           Madrid            1.46   
1           Centro                796           Madrid            1.03   
2           Centro                796           Madrid            0.59   
3           Centro                796           Madrid            0.75   
4           Centro                796           Madrid            0.93   
..              ...               ...              ...             ...   
123        Barajas                796           Madrid            1.98   
124        Barajas                796           Madrid           19.87   
125        Barajas                796           Madrid            0.64   
126        Barajas                796           Madrid           16.40   
127        Barajas                796           Madrid            4.67   

     densidad_por_km2  
0            15323.29  
1            43345.63  
2            17850.85  
3            21866.67  
4            33051.61  
..                ...  
123           9795.45  
124             89.88  
125          11415.62  
126            672.01  
127           1577.09  

[128 rows x 8 columns]

Hasta aquí el DF de los 131 barrios de Madrid (realizada copia renombrado cómodo), limpiado del indice, usado como índice el Codigo de Distrito

------------

### Conclusiones

Observamos que tenemos 915 registros (pisos), con 13 columnas, entre las que observamos:

- Url y Listing URL parecen "duplicadas". En todo caso, para nuestro estudio no tienen más relevancia que poder verificar, si fuera el caso, el origen y veracidad de los datos que tenemos. 
- Las dos últimas, advertiserProfessionalName y advertiserName, de nuevo, parecen duplicadas y carecen de interes a priori
- La columna id constituye un indice y puede funcionar como tal, eliminando la columna 0 de indice del dataframe
- nombres de columnas en inglés, que consideraré si me es más operativo traducir
- me llama la atención que title (al igual que Adress, Tipology, etc) no sea una cadena de texto sino un objeto
- observo que en 'description' hay un valor único que habrá que estudiar
- el número de 'Address' unicos es 105, lo cual no son los 131 distritos de Madrid. Pero sí hemos comprobado que los 105 valores únicos de Address son todos nombres de distritos de la ciudad de Madrid.




Madrid tiene 21 distritos y 131 barrios. Aunque es este DF no tenemos ninguno de ambas columnas, sería interesante poder disponer al menos de la columna distrito para cada inmueble, para poder hacer hipótesis y análisis.

![](Mapa-Madrid-Distritos-Barrios.jpg)

En este momento, y por cuestiones de visualización de las columnas en el ancho de la pantalla, voy a eliminar las columnas indicadas como duplicadas, listingUrl, y advertiserProfessionalName. Así nos quedaremos con 11 columnas


In [17]:
df.drop('listingUrl', axis=1, inplace=True, )
df.drop('advertiserProfessionalName', axis=1, inplace=True)
df

url  \
0    https://www.idealista.com/inmueble/104027174/   
1    https://www.idealista.com/inmueble/102321942/   
2    https://www.idealista.com/inmueble/103334142/   
3    https://www.idealista.com/inmueble/104161987/   
4    https://www.idealista.com/inmueble/103989666/   
..                                             ...   
910  https://www.idealista.com/inmueble/103960809/   
911  https://www.idealista.com/inmueble/103494183/   
912  https://www.idealista.com/inmueble/102107584/   
913   https://www.idealista.com/inmueble/99991709/   
914  https://www.idealista.com/inmueble/103315777/   

                                         title         id    price  baths  \
0         Piso en venta en calle de Villanueva  104027174  1920000      3   
1    Piso en venta en calle de Núñez de Balboa  102321942  1995000      3   
2         Piso en venta en Conde Orgaz-Piovera  103334142  1300000      3   
3                Piso en venta en Nueva España  104161987  1650000      3   
4     Piso en venta en calle de Claudio Coello  103989666  1590750      3   
..                                         ...        ...      ...    ...   
910       Piso en venta en calle de San Germán  103960809  1475000      4   
911       Piso en venta en calle de Fuencarral  103494183   865000      2   
912                Piso en venta en Castellana  102107584  7900000      5   
913               Piso en venta en Costillares   99991709   990000      4   
914            Piso en venta en Cuatro Caminos  103315777  1080000      2   

     rooms  sqft                                        description  \
0        3   183  Residencia única con acabados de la más alta c...   
1        3   170  Preciosa reforma a estrenar, con terrazas y ga...   
2        4   270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
3        3   248  Gilmar Real Estate vende espectacular piso en ...   
4        3   116  Descubre esta impresionante vivienda exterior ...   
..     ...   ...                                                ...   
910      5   376  Exclusiva vivienda en San GermánEngel & Völker...   
911      2   127  Esta magnífica vivienda, situada en una emblem...   
912      4   473  Álvora Capital Properties, consultora inmobili...   
913      6   293  Espacioso y tranquilo piso en Pinar de Chamart...   
914      4   270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                         address typology                      advertiserName  
0              Recoletos, Madrid    Pisos                      Promora Madrid  
1             Castellana, Madrid    Pisos              Engel & Völkers Madrid  
2              Hortaleza, Madrid    Pisos  De Salas Consultores Inmobiliarios  
3              Chamartín, Madrid    Pisos             Gilmar Viso - Chamartín  
4              Recoletos, Madrid    Pisos                  Walter Haus Madrid  
..                           ...      ...                                 ...  
910    Cuzco-Castillejos, Madrid    Pisos              Engel & Völkers Madrid  
911            Trafalgar, Madrid    Pisos                  Walter Haus Madrid  
912  Barrio de Salamanca, Madrid    Pisos           álvora capital properties  
913        Ciudad Lineal, Madrid    Pisos              Engel & Völkers Madrid  
914               Tetuán, Madrid    Pisos      Única Consultoría Inmobiliaria  

[915 rows x 11 columns]

Comprobado que los valores de ID son únicos, vamos a eliminar el índice y ussar como índice esta columna.

In [18]:
df = df.reset_index(drop=True).set_index("id")
df

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

           sqft                                        description  \
id                                                                   
104027174   183  Residencia única con acabados de la más alta c...   
102321942   170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142   270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987   248  Gilmar Real Estate vende espectacular piso en ...   
103989666   116  Descubre esta impresionante vivienda exterior ...   
...         ...                                                ...   
103960809   376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183   127  Esta magnífica vivienda, situada en una emblem...   
102107584   473  Álvora Capital Properties, consultora inmobili...   
99991709    293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777   270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                               address typology  \
id                                                
104027174            Recoletos, Madrid    Pisos   
102321942           Castellana, Madrid    Pisos   
103334142            Hortaleza, Madrid    Pisos   
104161987            Chamartín, Madrid    Pisos   
103989666            Recoletos, Madrid    Pisos   
...                                ...      ...   
103960809    Cuzco-Castillejos, Madrid    Pisos   
103494183            Trafalgar, Madrid    Pisos   
102107584  Barrio de Salamanca, Madrid    Pisos   
99991709         Ciudad Lineal, Madrid    Pisos   
103315777               Tetuán, Madrid    Pisos   

                               advertiserName  
id                                             
104027174                      Promora Madrid  
102321942              Engel & Völkers Madrid  
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
103989666                  Walter Haus Madrid  
...                                       ...  
103960809              Engel & Völkers Madrid  
103494183                  Walter Haus Madrid  
102107584           álvora capital properties  
99991709               Engel & Völkers Madrid  
103315777      Única Consultoría Inmobiliaria  

[915 rows x 10 columns]

Vamos a investigar ese valor nulo que aparece en descripción, a qué se debe.

In [19]:
fila_nulo_desc = df[df['description'].isnull()]
fila_nulo_desc 

url  \
id                                                         
103911168  https://www.idealista.com/inmueble/103911168/   

                              title   price  baths  rooms  sqft description  \
id                                                                            
103911168  Piso en venta en Palacio  260000      1      1    60         NaN   

                  address typology advertiserName  
id                                                 
103911168  Centro, Madrid    Pisos  Gilmar Centro

Dado que se trata de un único nulo y no es especialmente significativo ni por el tipo de inmueble o características, podríamos, bien eliminarlo, o bien sustituir el NaN por "No se aporta". De esta forma evitamos errores al trabajar con esa columna.

In [20]:
df.fillna('No se aporta', inplace=True)
df.loc[103911168]

url               https://www.idealista.com/inmueble/103911168/
title                                  Piso en venta en Palacio
price                                                    260000
baths                                                         1
rooms                                                         1
sqft                                                         60
description                                        No se aporta
address                                          Centro, Madrid
typology                                                  Pisos
advertiserName                                    Gilmar Centro
Name: 103911168, dtype: object

Vamos a transformar los datos de las columnas que contienen texto y que sin embargo son 'object'

In [21]:
print("Tipos de datos antes de la conversión:")
print(df.dtypes)

# Convertir columnas de tipo object a string
df["url"] = df["url"].astype("string")
df["title"] = df["title"].astype("string")
df["description"] = df["description"].astype("string")
df["address"] = df["address"].astype("string")
df["typology"] = df["typology"].astype("string")
df["advertiserName"] = df["advertiserName"].astype("string")

# Convertir columnas a int16
df["baths"] = df["baths"].astype("int16")
df["rooms"] = df["rooms"].astype("int16")
df["sqft"] = df["sqft"].astype("int16")


print("\nTipos de datos después de la conversión:")
print(df.dtypes)

Tipos de datos antes de la conversión:
url               object
title             object
price              int64
baths              int64
rooms              int64
sqft               int64
description       object
address           object
typology          object
advertiserName    object
dtype: object

Tipos de datos después de la conversión:
url               string[python]
title             string[python]
price                      int64
baths                      int16
rooms                      int16
sqft                       int16
description       string[python]
address           string[python]
typology          string[python]
advertiserName    string[python]
dtype: object


In [22]:
df.memory_usage()

Index             40384
url                7320
title              7320
price              7320
baths              1830
rooms              1830
sqft               1830
description        7320
address            7320
typology           7320
advertiserName     7320
dtype: int64

Dado que no soporto que la columna que contiene metros cuadrados se llame pies cuadrados 'sqft', voy a renombrarla. Lo cierto es que no tengo forma de comprobar que el contenido de dicha columna son m2, pero dado que los daatos son de pisos en España, ex seguro que es así. 

In [23]:
df.rename(columns={'sqft': 'm2'}, inplace= True)

#y estas dos de paso, para facilitar luego trabajar con ellas
df.rename(columns={'typology': 'type', 'advertiserName': 'adName'}, inplace= True)
df

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                               address   type  \
id                                              
104027174            Recoletos, Madrid  Pisos   
102321942           Castellana, Madrid  Pisos   
103334142            Hortaleza, Madrid  Pisos   
104161987            Chamartín, Madrid  Pisos   
103989666            Recoletos, Madrid  Pisos   
...                                ...    ...   
103960809    Cuzco-Castillejos, Madrid  Pisos   
103494183            Trafalgar, Madrid  Pisos   
102107584  Barrio de Salamanca, Madrid  Pisos   
99991709         Ciudad Lineal, Madrid  Pisos   
103315777               Tetuán, Madrid  Pisos   

                                       adName  
id                                             
104027174                      Promora Madrid  
102321942              Engel & Völkers Madrid  
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
103989666                  Walter Haus Madrid  
...                                       ...  
103960809              Engel & Völkers Madrid  
103494183                  Walter Haus Madrid  
102107584           álvora capital properties  
99991709               Engel & Völkers Madrid  
103315777      Única Consultoría Inmobiliaria  

[915 rows x 10 columns]

Con lo que tenemos hasta aquí, vamos a limpiar la columna de address, eliminando el nombre de la ciudad, y por lo tanto, asumiédola como barrio, a la vez que incorporamos una columna de distrito, obtenida de las tablas anteriores.


In [24]:
# Limpiar la columna 'address'
df['address'] = df['address'].str.split(',').str[0]

df

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                       address   type                              adName  
id                                                                         
104027174            Recoletos  Pisos                      Promora Madrid  
102321942           Castellana  Pisos              Engel & Völkers Madrid  
103334142            Hortaleza  Pisos  De Salas Consultores Inmobiliarios  
104161987            Chamartín  Pisos             Gilmar Viso - Chamartín  
103989666            Recoletos  Pisos                  Walter Haus Madrid  
...                        ...    ...                                 ...  
103960809    Cuzco-Castillejos  Pisos              Engel & Völkers Madrid  
103494183            Trafalgar  Pisos                  Walter Haus Madrid  
102107584  Barrio de Salamanca  Pisos           álvora capital properties  
99991709         Ciudad Lineal  Pisos              Engel & Völkers Madrid  
103315777               Tetuán  Pisos      Única Consultoría Inmobiliaria  

[915 rows x 10 columns]

Algunas observaciones de las direcciones:
- El distrito Palos de Moguer a pasado a llamarse Palos de la Frontera, renombro esos campos para que coincida
- Sanchinarro es un area residencial del barrio de Valdefuentes
- Virgen del Cortijo es un area residencial del barrio de Valdefuentes.
-c orrijo algunos nombres mal escritos


In [25]:
# Renombrar varios valores a la vez
# este renombrado viene de buscar los valores raros, normalmente en mapa y wikipedia, y ubicarlos
renombrar = {
    "Palos de Moguer": "Palos de la Frontera",
    "Barrio de Salamanca": "Salamanca",
    "Sanchinarro":"Valdefuentes", 
    "Buena Vista":"Buenavista",
    "Virgen del Cortijo - Manoteras": "Valdefuentes",
    "Pacifico":"Pacífico"
}

df['address'] = df['address'].replace(renombrar)

df

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                     address   type                              adName  
id                                                                       
104027174          Recoletos  Pisos                      Promora Madrid  
102321942         Castellana  Pisos              Engel & Völkers Madrid  
103334142          Hortaleza  Pisos  De Salas Consultores Inmobiliarios  
104161987          Chamartín  Pisos             Gilmar Viso - Chamartín  
103989666          Recoletos  Pisos                  Walter Haus Madrid  
...                      ...    ...                                 ...  
103960809  Cuzco-Castillejos  Pisos              Engel & Völkers Madrid  
103494183          Trafalgar  Pisos                  Walter Haus Madrid  
102107584          Salamanca  Pisos           álvora capital properties  
99991709       Ciudad Lineal  Pisos              Engel & Völkers Madrid  
103315777             Tetuán  Pisos      Única Consultoría Inmobiliaria  

[915 rows x 10 columns]

Vamos a intentar comprobar si todos los datos que figuran en la columna address, realmente son un barrio de la ciudad de Madrid, según la tabla barrios_Madrid, obtenida de fuentes oficiales de la administración.

In [26]:
# Comprobar si todos los valores de 'address' están en 'nombre'
coincidentes = df['address'].isin(bM['NOMBRE']).all()

# Mostrar resultado
if coincidentes:
    print("Todos los valores de 'address' están presentes en 'nombre'.")
else:
    print("No todos los valores de 'address' están presentes en 'nombre'.")

No todos los valores de 'address' están presentes en 'nombre'.


In [27]:
valores_faltantes_barrio = df[~df['address'].isin(bM['NOMBRE'])]
print("Filas cuya 'address' que no es un barrio de Madrid:")
valores_faltantes_barrio


Filas cuya 'address' que no es un barrio de Madrid:


url  \
id                                                         
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
101582665  https://www.idealista.com/inmueble/101582665/   
102164326  https://www.idealista.com/inmueble/102164326/   
103226676  https://www.idealista.com/inmueble/103226676/   
...                                                  ...   
103392553  https://www.idealista.com/inmueble/103392553/   
103960809  https://www.idealista.com/inmueble/103960809/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                                      title    price  baths  \
id                                                                            
103334142              Piso en venta en Conde Orgaz-Piovera  1300000      3   
104161987                     Piso en venta en Nueva España  1650000      3   
101582665  Chalet adosado en venta en paseo de los Jacintos   995000      3   
102164326             Piso en venta en calle de las Huertas  1175000      3   
103226676    Piso en venta en Nuevos Ministerios-Ríos Rosas  1075000      3   
...                                                     ...      ...    ...   
103392553                          Piso en venta en Palacio   500000      2   
103960809              Piso en venta en calle de San Germán  1475000      4   
102107584                       Piso en venta en Castellana  7900000      5   
99991709                       Piso en venta en Costillares   990000      4   
103315777                   Piso en venta en Cuatro Caminos  1080000      2   

           rooms   m2                                        description  \
id                                                                         
103334142      4  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987      3  248  Gilmar Real Estate vende espectacular piso en ...   
101582665      3  160  Chalet adosado luminoso en HispanámericaEngel ...   
102164326      3  178  Amplio y clásico piso en Huertas-CortesEngel&V...   
103226676      3  169  Luminoso piso con terraza en ChamberíEngel & V...   
...          ...  ...                                                ...   
103392553      2   99  Elegante y moderna propiedad en La Latina, Pal...   
103960809      5  376  Exclusiva vivienda en San GermánEngel & Völker...   
102107584      4  473  Álvora Capital Properties, consultora inmobili...   
99991709       6  293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777      4  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                           address            type  \
id                                                   
103334142                Hortaleza           Pisos   
104161987                Chamartín           Pisos   
101582665  Bernabéu-Hispanoamérica  Independientes   
102164326           Huertas-Cortes           Pisos   
103226676                 Chamberí           Pisos   
...                            ...             ...   
103392553                   Centro           Pisos   
103960809        Cuzco-Castillejos           Pisos   
102107584                Salamanca           Pisos   
99991709             Ciudad Lineal           Pisos   
103315777                   Tetuán           Pisos   

                                       adName  
id                                             
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
101582665              Engel & Völkers Madrid  
102164326              Engel & Völkers Madrid  
103226676              Engel & Völkers Madrid  
...                                       ...  
103392553              Engel & Völkers Madrid  
103960809              Engel & Völkers Madrid  
102107584           álvora capital properties  
99991709               E

Algunas direcciones no son barrios de Madrid, sino directamente, distrito de Madrid. Voy a hacer una busqueda sobre lo anterior, para descartar también estos, y poder quedarme con los que no son ni una cosa ni otra... a ver qué son...

In [28]:
valores_faltantes_distrito = valores_faltantes_barrio[~valores_faltantes_barrio['address'].isin(dM['NOMBRE'])]
print("Filas cuya 'address' que no es un distrito de Madrid (ni tampoco era un barrio):")
valores_faltantes_distrito


Filas cuya 'address' que no es un distrito de Madrid (ni tampoco era un barrio):


url  \
id                                                         
101582665  https://www.idealista.com/inmueble/101582665/   
102164326  https://www.idealista.com/inmueble/102164326/   
103326278  https://www.idealista.com/inmueble/103326278/   
103741155  https://www.idealista.com/inmueble/103741155/   
104149846  https://www.idealista.com/inmueble/104149846/   
...                                                  ...   
102641390  https://www.idealista.com/inmueble/102641390/   
102123027  https://www.idealista.com/inmueble/102123027/   
102616450  https://www.idealista.com/inmueble/102616450/   
103692239  https://www.idealista.com/inmueble/103692239/   
103960809  https://www.idealista.com/inmueble/103960809/   

                                                      title    price  baths  \
id                                                                            
101582665  Chalet adosado en venta en paseo de los Jacintos   995000      3   
102164326             Piso en venta en calle de las Huertas  1175000      3   
103326278         Piso en venta en avenida de Concha Espina  1970000      4   
103741155           Piso en venta en calle de Cervantes, 22  1800000      4   
104149846   Chalet pareado en venta en calle de el Algabeño  1650000      5   
...                                                     ...      ...    ...   
102641390                      Ático en venta en Valdemarín  1540000      3   
102123027              Dúplex en venta en plaza De Lavapies   290000      2   
102616450   Chalet adosado en venta en calle José Rizal s/n  1650000      5   
103692239        Piso en venta en calle de Manuela Malasaña  1150000      3   
103960809              Piso en venta en calle de San Germán  1475000      4   

           rooms   m2                                        description  \
id                                                                         
101582665      3  160  Chalet adosado luminoso en HispanámericaEngel ...   
102164326      3  178  Amplio y clásico piso en Huertas-CortesEngel&V...   
103326278      4  283  PROMORA MADRID vende precioso piso en barrio C...   
103741155      3  214  DIZA Consultores les presenta excepcional piso...   
104149846      5  367  Suma inmobiliaria ubicada en la prestigiosa ur...   
...          ...  ...                                                ...   
102641390      4  345  Ático dúplex de 345 m² con una magnífica terra...   
102123027      2   62  ¡Descubre tu oportunidad excepcional en Gilmar...   
102616450      5  360  QUALITÉ Agence Immobilière Française présente ...   
103692239      3  133  PROMORA MADRID vende piso reformado en zona Ce...   
103960809      5  376  Exclusiva vivienda en San GermánEngel & Völker...   

                           address            type  \
id                                                   
101582665  Bernabéu-Hispanoamérica  Independientes   
102164326           Huertas-Cortes           Pisos   
103326278  Bernabéu-Hispanoamérica           Pisos   
103741155           Huertas-Cortes           Pisos   
104149846      Conde Orgaz-Piovera  Independientes   
...                            ...             ...   
102641390                  Moncloa           Pisos   
102123027     Lavapiés-Embajadores           Pisos   
102616450      Conde Orgaz-Piovera  Independientes   
103692239     Malasaña-Universidad           Pisos   
103960809        Cuzco-Castillejos           Pisos   

                                                      adName  
id                                                            
101582665                             Engel & Völkers Madrid  
102164326                             Engel & Völkers Madrid  
103326278                                     Promora Madrid  
103741155  DIZA Consultores Inmobiliaria - Propiedades de...  
104149846                                  SUMA Inmobiliaria  
...                                                      ...  
102641390                                    Pro

Aún tengo 292 apariciones así que voy a refinar la selección, si contienen una palabra que sea barrio o distrito.

In [29]:
# Función para verificar coincidencias parciales de 'adress' con los nombres de los barrios
def verificar_coincidencia(address, barrios):
    for barrio in barrios:
        if barrio in address or address in barrio:
            return True
    return False

# Aplicar la función a cada valor de 'address'
barrios_lista = bM['NOMBRE'].tolist()  # Convertir la columna barrios a lista
df['coincide'] = df['address'].apply(lambda x: verificar_coincidencia(x, barrios_lista))

# Comprobar si todos los valores tienen coincidencia
todos_coinciden = df['coincide'].all()

# Resultados
if todos_coinciden:
    print("Todos los valores de 'address' tienen coincidencia parcial con algún barrio.")
else:
    print("Algunos valores de 'address' no tienen coincidencia parcial con ningún barrio.")

Algunos valores de 'address' no tienen coincidencia parcial con ningún barrio.


In [30]:
valores_faltantes_barrio = df[~df['coincide']]
print("Filas cuya 'address' que no incluyen en su texto un barrio de Madrid:")
valores_faltantes_barrio

Filas cuya 'address' que no incluyen en su texto un barrio de Madrid:


url  \
id                                                         
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103226676  https://www.idealista.com/inmueble/103226676/   
104079750  https://www.idealista.com/inmueble/104079750/   
103725861  https://www.idealista.com/inmueble/103725861/   
...                                                  ...   
103369764  https://www.idealista.com/inmueble/103369764/   
103392553  https://www.idealista.com/inmueble/103392553/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                                       title    price  baths  \
id                                                                             
103334142               Piso en venta en Conde Orgaz-Piovera  1300000      3   
104161987                      Piso en venta en Nueva España  1650000      3   
103226676     Piso en venta en Nuevos Ministerios-Ríos Rosas  1075000      3   
104079750                Chalet pareado en venta en Canillas  1850000      5   
103725861                    Piso en venta en Huertas-Cortes   780000      2   
...                                                      ...      ...    ...   
103369764  Casa o chalet independiente en venta en Urb. V...   870000      3   
103392553                           Piso en venta en Palacio   500000      2   
102107584                        Piso en venta en Castellana  7900000      5   
99991709                        Piso en venta en Costillares   990000      4   
103315777                    Piso en venta en Cuatro Caminos  1080000      2   

           rooms   m2                                        description  \
id                                                                         
103334142      4  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987      3  248  Gilmar Real Estate vende espectacular piso en ...   
103226676      3  169  Luminoso piso con terraza en ChamberíEngel & V...   
104079750      5  438  Suma inmobiliaria ofrece, EN EXCLUSIVA, este c...   
103725861      2  117  PISO EXTERIOR EN EDIFICIO REPRESENTATIVO CON B...   
...          ...  ...                                                ...   
103369764      3  314  GILMAR Consulting Inmobiliario Conde Orgaz, co...   
103392553      2   99  Elegante y moderna propiedad en La Latina, Pal...   
102107584      4  473  Álvora Capital Properties, consultora inmobili...   
99991709       6  293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777      4  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                 address            type  \
id                                         
103334142      Hortaleza           Pisos   
104161987      Chamartín           Pisos   
103226676       Chamberí           Pisos   
104079750      Hortaleza  Independientes   
103725861         Centro           Pisos   
...                  ...             ...   
103369764      Hortaleza  Independientes   
103392553         Centro           Pisos   
102107584      Salamanca           Pisos   
99991709   Ciudad Lineal           Pisos   
103315777         Tetuán           Pisos   

                                             adName  coincide  
id                                                             
103334142        De Salas Consultores Inmobiliarios     False  
104161987                   Gilmar Viso - Chamartín     False  
103226676                    Engel & Völkers Madrid     False  
104079750                         SUMA Inmobiliaria     False  
103725861            aProperties Real Estate Madrid     False  
...                                             ...       ...  
103369764  gilmar conde orgaz-campo de las naciones     False  
103392553                    Engel & Völkers Madrid     False  
102107584                 álv

In [31]:
# Función para verificar coincidencias parciales de 'address' con los nombres de los distritos
def verificar_coincidencia(address, distritos):
    for distrito in distritos:
        if distrito in address or address in distrito:
            return True
    return False

# Aplicar la función a cada valor de 'address'
distritos_lista = dM['NOMBRE'].tolist()  # Convertir la columna distritos a lista
valores_faltantes_barrio['coincide'] = valores_faltantes_barrio['address'].apply(lambda x: verificar_coincidencia(x, distritos_lista))

# Comprobar si todos los valores tienen coincidencia
todos_coinciden = valores_faltantes_barrio['coincide'].all()

# Resultados
if todos_coinciden:
    print("Todos los valores de 'address' tienen coincidencia parcial con algún distrito.")
else:
    print("Algunos valores de 'address' no tienen coincidencia parcial con ningún distrito.")

Todos los valores de 'address' tienen coincidencia parcial con algún distrito.


/var/folders/lh/mx1_xdkx76s1jt00m2qwl2km0000gn/T/ipykernel_89369/3812476530.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  valores_faltantes_barrio['coincide'] = valores_faltantes_barrio['address'].apply(lambda x: verificar_coincidencia(x, distritos_lista))


In [32]:
valores_faltantes_distrito = valores_faltantes_barrio[~valores_faltantes_barrio['coincide']]
print("Filas cuya 'address' que no incluyen en su texto un distrito de Madrid (y tampoco un barrio):")
valores_faltantes_distrito


Filas cuya 'address' que no incluyen en su texto un distrito de Madrid (y tampoco un barrio):


Empty DataFrame
Columns: [url, title, price, baths, rooms, m2, description, address, type, adName, coincide]
Index: []

Así pues, ahora podemos incorporar, segregar las columnas distrito y barrio. Antes, elimino la columna última de 'Coincide'

In [33]:
df = df.drop ('coincide',axis=1)

In [34]:
#en este punto, por precaución, hago una copia del df y creo uno procesado, df2
df2 = df.copy()
df2


url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                     address   type                              adName  
id                                                                       
104027174          Recoletos  Pisos                      Promora Madrid  
102321942         Castellana  Pisos              Engel & Völkers Madrid  
103334142          Hortaleza  Pisos  De Salas Consultores Inmobiliarios  
104161987          Chamartín  Pisos             Gilmar Viso - Chamartín  
103989666          Recoletos  Pisos                  Walter Haus Madrid  
...                      ...    ...                                 ...  
103960809  Cuzco-Castillejos  Pisos              Engel & Völkers Madrid  
103494183          Trafalgar  Pisos                  Walter Haus Madrid  
102107584          Salamanca  Pisos           álvora capital properties  
99991709       Ciudad Lineal  Pisos              Engel & Völkers Madrid  
103315777             Tetuán  Pisos      Única Consultoría Inmobiliaria  

[915 rows x 10 columns]

In [39]:
# Función para encontrar coincidencia de address con barrio e incorporarlo a una nueva columna 
# en caso de no existir, colocaremos 'ND'
def obtener_barrio(address, barrios):
    for barrio in barrios:
        if barrio in address or address in barrio:
            return barrio
    return "ND"

# Crear nueva columna con el resultado
barrios_lista = bM['NOMBRE'].tolist()
df2['Barrio'] = df2['address'].apply(lambda x: obtener_barrio(x, barrios_lista))

# Mover la columna justo despues de Address
col = df2.pop('Barrio')
df2.insert (9, 'Barrio',col)

df2["Barrio"] = df2["Barrio"].astype("string")
df2

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                     address   type       Barrio  \
id                                                 
104027174          Recoletos  Pisos    Recoletos   
102321942         Castellana  Pisos   Castellana   
103334142          Hortaleza  Pisos           ND   
104161987          Chamartín  Pisos           ND   
103989666          Recoletos  Pisos    Recoletos   
...                      ...    ...          ...   
103960809  Cuzco-Castillejos  Pisos  Castillejos   
103494183          Trafalgar  Pisos    Trafalgar   
102107584          Salamanca  Pisos           ND   
99991709       Ciudad Lineal  Pisos           ND   
103315777             Tetuán  Pisos           ND   

                                       adName  
id                                             
104027174                      Promora Madrid  
102321942              Engel & Völkers Madrid  
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
103989666                  Walter Haus Madrid  
...                                       ...  
103960809              Engel & Völkers Madrid  
103494183                  Walter Haus Madrid  
102107584           álvora capital properties  
99991709               Engel & Völkers Madrid  
103315777      Única Consultoría Inmobiliaria  

[915 rows x 11 columns]

Hago lo mismo si coincide con el nombre de distrito y lo incorporo en una nueva columna

In [41]:
# Función para encontrar coincidencia de address con distritos e incorporarlo a una nueva columna 
# y en el caso de no coincidir, buscaremos aqué distrito pertenece ese barrio
def obtener_distrito(address, distritos):
    for distrito in distritos:
        if distrito in address or address in distrito:
            return distrito
    return "completar"

# Crear nueva columna con el resultado
distritos_lista = dM['NOMBRE'].tolist()
df2['Distrito'] = df2['address'].apply(lambda x: obtener_distrito(x, distritos_lista))

# Mover la columna justo despues de Address
col = df2.pop('Distrito')
df2.insert (10, 'Distrito',col)

df2["Distrito"] = df2["Distrito"].astype("string")

df2

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                     address   type       Barrio       Distrito  \
id                                                                
104027174          Recoletos  Pisos    Recoletos      completar   
102321942         Castellana  Pisos   Castellana      completar   
103334142          Hortaleza  Pisos           ND      Hortaleza   
104161987          Chamartín  Pisos           ND      Chamartín   
103989666          Recoletos  Pisos    Recoletos      completar   
...                      ...    ...          ...            ...   
103960809  Cuzco-Castillejos  Pisos  Castillejos      completar   
103494183          Trafalgar  Pisos    Trafalgar      completar   
102107584          Salamanca  Pisos           ND      Salamanca   
99991709       Ciudad Lineal  Pisos           ND  Ciudad Lineal   
103315777             Tetuán  Pisos           ND         Tetuán   

                                       adName  
id                                             
104027174                      Promora Madrid  
102321942              Engel & Völkers Madrid  
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
103989666                  Walter Haus Madrid  
...                                       ...  
103960809              Engel & Völkers Madrid  
103494183                  Walter Haus Mad

In [42]:
# Función para buscar el distrito correspondiente al barrio en bM
def obtener_distrito(barrio, bM):
    # Buscar el distrito en el dataframe bM
    distrito = bM.loc[bM['NOMBRE'] == barrio, 'NOMDIS']
    if not distrito.empty:
        return distrito.iloc[0]
    return None  # Si no se encuentra, devolver None

# Aplicar la función solo a las filas donde 'distrito' está vacío
df2['Distrito'] = df2.apply(
    lambda row: obtener_distrito(row['Barrio'], bM) if row['Distrito'] == "completar" else row['Distrito'],
    axis=1
)

df2

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description  \
id                                                                  
104027174  183  Residencia única con acabados de la más alta c...   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...   
104161987  248  Gilmar Real Estate vende espectacular piso en ...   
103989666  116  Descubre esta impresionante vivienda exterior ...   
...        ...                                                ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...   
103494183  127  Esta magnífica vivienda, situada en una emblem...   
102107584  473  Álvora Capital Properties, consultora inmobili...   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...   

                     address   type       Barrio       Distrito  \
id                                                                
104027174          Recoletos  Pisos    Recoletos      Salamanca   
102321942         Castellana  Pisos   Castellana      Salamanca   
103334142          Hortaleza  Pisos           ND      Hortaleza   
104161987          Chamartín  Pisos           ND      Chamartín   
103989666          Recoletos  Pisos    Recoletos      Salamanca   
...                      ...    ...          ...            ...   
103960809  Cuzco-Castillejos  Pisos  Castillejos         Tetuán   
103494183          Trafalgar  Pisos    Trafalgar       Chamberí   
102107584          Salamanca  Pisos           ND      Salamanca   
99991709       Ciudad Lineal  Pisos           ND  Ciudad Lineal   
103315777             Tetuán  Pisos           ND         Tetuán   

                                       adName  
id                                             
104027174                      Promora Madrid  
102321942              Engel & Völkers Madrid  
103334142  De Salas Consultores Inmobiliarios  
104161987             Gilmar Viso - Chamartín  
103989666                  Walter Haus Madrid  
...                                       ...  
103960809              Engel & Völkers Madrid  
103494183                  Walter Haus Mad

In [43]:
df2.drop ('address',axis=1, inplace=True) #eliminamos ya la columna address, cuyo contenido hemos desdoblado y trabajado.
df2

url  \
id                                                         
104027174  https://www.idealista.com/inmueble/104027174/   
102321942  https://www.idealista.com/inmueble/102321942/   
103334142  https://www.idealista.com/inmueble/103334142/   
104161987  https://www.idealista.com/inmueble/104161987/   
103989666  https://www.idealista.com/inmueble/103989666/   
...                                                  ...   
103960809  https://www.idealista.com/inmueble/103960809/   
103494183  https://www.idealista.com/inmueble/103494183/   
102107584  https://www.idealista.com/inmueble/102107584/   
99991709    https://www.idealista.com/inmueble/99991709/   
103315777  https://www.idealista.com/inmueble/103315777/   

                                               title    price  baths  rooms  \
id                                                                            
104027174       Piso en venta en calle de Villanueva  1920000      3      3   
102321942  Piso en venta en calle de Núñez de Balboa  1995000      3      3   
103334142       Piso en venta en Conde Orgaz-Piovera  1300000      3      4   
104161987              Piso en venta en Nueva España  1650000      3      3   
103989666   Piso en venta en calle de Claudio Coello  1590750      3      3   
...                                              ...      ...    ...    ...   
103960809       Piso en venta en calle de San Germán  1475000      4      5   
103494183       Piso en venta en calle de Fuencarral   865000      2      2   
102107584                Piso en venta en Castellana  7900000      5      4   
99991709                Piso en venta en Costillares   990000      4      6   
103315777            Piso en venta en Cuatro Caminos  1080000      2      4   

            m2                                        description   type  \
id                                                                         
104027174  183  Residencia única con acabados de la más alta c...  Pisos   
102321942  170  Preciosa reforma a estrenar, con terrazas y ga...  Pisos   
103334142  270  DE SALAS CONSULTORES INMOBILIARIOS MAV02679 le...  Pisos   
104161987  248  Gilmar Real Estate vende espectacular piso en ...  Pisos   
103989666  116  Descubre esta impresionante vivienda exterior ...  Pisos   
...        ...                                                ...    ...   
103960809  376  Exclusiva vivienda en San GermánEngel & Völker...  Pisos   
103494183  127  Esta magnífica vivienda, situada en una emblem...  Pisos   
102107584  473  Álvora Capital Properties, consultora inmobili...  Pisos   
99991709   293  Espacioso y tranquilo piso en Pinar de Chamart...  Pisos   
103315777  270  Única Inmobiliaria Chamartín ofrece esta gran ...  Pisos   

                Barrio       Distrito                              adName  
id                                                                         
104027174    Recoletos      Salamanca                      Promora Madrid  
102321942   Castellana      Salamanca              Engel & Völkers Madrid  
103334142           ND      Hortaleza  De Salas Consultores Inmobiliarios  
104161987           ND      Chamartín             Gilmar Viso - Chamartín  
103989666    Recoletos      Salamanca                  Walter Haus Madrid  
...                ...            ...                                 ...  
103960809  Castillejos         Tetuán              Engel & Völkers Madrid  
103494183    Trafalgar       Chamberí                  Walter Haus Madrid  
102107584           ND      Salamanca           álvora capital properties  
99991709            ND  Ciudad Lineal              Engel & Völkers Madrid  
103315777           ND         Tetuán      Única Consultoría Inmobiliaria  

[915 rows x 11 columns]

Nos quedamos con esta matriz de 11 columnas, con información de barrio y distrito.

Como última operación, dado que viene en el dataframe, se incorpora una columna con el precios/m2, resultante de ambas columnas.

In [44]:
df2 ['euros_m2'] = round(df2 ['price'] / df2 ['m2'],2)

# Mover la columna justo despues de titleAddress
col = df2.pop('euros_m2')
df2.insert (2, 'euros_m2',col)


In [53]:
# Guardar el DataFrame en un archivo JSON
df2.to_json("../data/processed/viviendas_EDA_agv.json", indent=4, index = True)

print("El archivo JSON se ha guardado como 'viviendas_EDA_agv.json' en la carpeta 'data/processed'.")

El archivo JSON se ha guardado como 'viviendas_EDA_agv.json' en la carpeta 'data/processed'.


In [50]:
# Guardar el DataFrame en un archivo CSV
df2.to_csv("../data/processed/viviendas_EDA_agv.csv", index = True)

print("El archivo CSV se ha guardado como 'viviendas_EDA_agv.csv' en la carpeta 'data/processed'.")

El archivo CSV se ha guardado como 'viviendas_EDA_agv.csv' en la carpeta 'data/processed'.


In [47]:
df2.describe()

euros_m2         price       baths       rooms           m2
count    915.000000  9.150000e+02  915.000000  915.000000   915.000000
mean    6400.946022  1.290575e+06    2.900546    3.474317   217.238251
std     2800.261853  1.036640e+06    1.626487    1.594625   191.745826
min     1198.080000  1.220000e+05    1.000000    0.000000    14.000000
25%     4394.450000  6.250000e+05    2.000000    2.000000   104.000000
50%     5760.870000  9.900000e+05    3.000000    3.000000   158.000000
75%     7937.335000  1.650000e+06    4.000000    4.000000   264.000000
max    21299.640000  8.900000e+06   12.000000   12.000000  1768.000000